CMakeLists.txt

```diff
@@ -52,27 +52,135 @@ target_link_libraries(logging_client PRIVATE
   http_parser.host
 )
 
-add_enclave_lib(loggingenc
-  SRCS
-    src/apps/logging/logging.cpp
-    src/apps/logging/stub_for_code_signing.cpp
-)
-sign_app_library(loggingenc
-  ${CMAKE_CURRENT_SOURCE_DIR}/src/apps/logging/oe_sign.conf
-  ${CMAKE_CURRENT_SOURCE_DIR}/src/apps/sample_key.pem
-)
+# Build common library for CCF enclaves
+add_custom_target(ccfcommon ALL)
+
+if("sgx" IN_LIST TARGET)
+  # enclave version
+  add_library(ccfcommon.enclave STATIC
+    ${CCF_DIR}/src/enclave/main.cpp
+    ${CCF_GENERATED_DIR}/ccf_t.cpp
+  )
+
+  target_compile_definitions(ccfcommon.enclave PUBLIC
+    INSIDE_ENCLAVE
+    _LIBCPP_HAS_THREAD_API_PTHREAD
+  )
+
+  target_compile_options(ccfcommon.enclave PUBLIC
+    -nostdinc
+    -nostdinc++
+  )
+
+  target_include_directories(ccfcommon.enclave SYSTEM PUBLIC
+    ${EVERCRYPT_INC}
+    ${CCF_GENERATED_DIR}
+  )
+
+  add_dependencies(ccfcommon.enclave flatbuffers)
+
+  if (PBFT)
+    target_link_libraries(ccfcommon.enclave PUBLIC
+      libbyz.enclave
+    )
+  endif()
+
+  target_link_libraries(ccfcommon.enclave PUBLIC
+    openenclave::oeenclave
+    openenclave::oelibcxx
+    ccfcrypto.enclave
+    evercrypt.enclave
+    http_parser.enclave
+    lua.enclave
+    secp256k1.enclave
+  )
+
+  set_property(TARGET ccfcommon.enclave
+    PROPERTY POSITION_INDEPENDENT_CODE ON
+  )
+
+  enable_quote_code(ccfcommon.enclave)
+
+  install(
+    TARGETS ccfcommon.enclave
+    DESTINATION lib
+  )
+
+  add_dependencies(ccfcommon ccfcommon.enclave)
+endif()
+
+if("virtual" IN_LIST TARGET)
+  # virtual version
+  add_library(ccfcommon.virtual STATIC
+    ${CCF_DIR}/src/enclave/main.cpp
+    ${CCF_GENERATED_DIR}/ccf_t.cpp
+  )
+
+  target_compile_definitions(ccfcommon.virtual PUBLIC
+    INSIDE_ENCLAVE
+    VIRTUAL_ENCLAVE
+    _LIBCPP_HAS_THREAD_API_PTHREAD
+  )
+
+  target_compile_options(ccfcommon.virtual PUBLIC
+    -stdlib=libc++
+  )
+
+  target_include_directories(ccfcommon.virtual SYSTEM PUBLIC
+    ${EVERCRYPT_INC}
+    ${CCF_GENERATED_DIR}
+    ${OE_INCLUDEDIR} # Virtual libraries don't link against OE, but do share includes
```

Comment:
Correct me if I'm wrong (I've just tested this locally) but I don't think we need this anymore? I don't think the two openenclave includes in `ccf_v.h` are still required.

Reply:
I've added some more stub definitions in `ccf_v.h` so this is now true. The key was removing `ccf_t.cpp` from the virtual build - this depends on `enclave.h`, but defines the things we're already stubbing. There's also now a preprocessor error if we try to include `ccf_v.h` with `GET_QUOTE` defined, rather than a run-time error and more empty typedefs.